src/NumericalAlgorithms/RootFinding/RootBracketing.hpp

```diff
@@ -0,0 +1,317 @@
+// containing points that have already been tried fro bracketing.
+//
+// Returns a std::array<double,4> containing {{x1,x2,y1,y2}} where
+// x1 and x2 bracket the root, and y1=f(x1) and y2=f(x2).
+//
+// Note that y might be undefined (i.e. an invalid boost::optional)
+// for some values of x.  But we assume that if f(x1) and f(x2) are
+// valid for some x1 and x2, then f(x) is valid for all x between x1
+// and x2.
+//
+// We also assume that there is only one root.
+//
+// So this means we have only 2 possibilities for the validity of the
+// points in the input x,y arrays:
+// 1) All points are invalid, e.g. "X X X X X X X X".
+//    (here X represents an invalid point)
+// 2) All valid points are adjacent, with the same sign, e.g. "X X + + X X"
+//    or "+ + X X X" or "X X X - -".
+//    (here + represents a point with y>0 and - represents a point with y<0).
+// Note that we assume that the input does not contain something
+// like "X X + + - - X X" because the caller would already know that the root
+// was bracketed in that case, and thus the caller would not call
+// bracket_by_contracting.
+//
+// For case 1) above, we bisect each pair of points, and call
+// bracket_by_contracting recursively until we find a valid point.
+// For case 2) above, it is sufficent to check for a bracket only
+// between valid and invalid points.  That is, for "X X + + X X" we
+// check only between points 1 and 2 and between points 3 and 4 (where
+// points are numbered starting from zero).  For "+ + X X X" we check
+// only between points 1 and 2.
+template <typename Functor>
+std::array<double, 4> bracket_by_contracting(
+    const std::vector<double>& x, const std::vector<boost::optional<double>>& y,
+    const Functor& f, const size_t level = 0) noexcept {
+  constexpr size_t max_level = 6;
+
+  // First check if we have any valid points.
+  size_t last_valid_index = y.size();
+  for (size_t i = y.size(); i >= 1; --i) {
+    if (y[i - 1]) {
+      last_valid_index = i - 1;
+      break;
+    }
+  }
+
+  if (last_valid_index == y.size()) {
+    // No valid points!
+
+    // Create larger arrays with one point between each of the already
+    // computed points.
+    std::vector<double> xx(x.size() * 2 - 1);
+    std::vector<boost::optional<double>> yy(y.size() * 2 - 1);
+
+    // Copy all even-numbered points in the range.
+    for (size_t i = 0; i < x.size(); ++i) {
+      xx[2 * i] = x[i];
+      yy[2 * i] = y[i];
+    }
+
+    // Fill midpoints and check for bracket on each one.
+    for (size_t i = 0; i < x.size() - 1; ++i) {
+      xx[2 * i + 1] = x[i] + 0.5 * (x[i + 1] - x[i]);
+      yy[2 * i + 1] = f(xx[2 * i + 1]);
+      if (yy[2 * i + 1]) {
+        // Valid point! We know that all the other points are
+        // invalid, so we need to check only 3 points in the next
+        // iteration: the new valid point and its neighbors.
+        if (level > max_level) {
+          ERROR("Too many iterations in bracket_by_contracting");
+        }
```

Comment:
I see this check being done in several different locations in the function. Why not replace these multiple checks with a single check on entering the function, at line 53ish?